m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+    private Phaser phaser = new Phaser(1);
+
+    // Use inner Builder class to construct an M3Reporter
+    private M3Reporter(Builder builder) {
+        try {
+            // Builder verifies non-null, non-empty hostPorts
+            String[] hostPorts = builder.hostPorts;
+
+            TProtocolFactory protocolFactory;
+            if (builder.protocol == ThriftProtocol.COMPACT) {
+                protocolFactory = new TCompactProtocol.Factory();
+            } else {
+                protocolFactory = new TBinaryProtocol.Factory();
+            }
+
+            TTransport transport = new TUdpClient(hostPorts[0]);
+
+            client = new M3.Client(protocolFactory.getProtocol(transport));
+            resourcePool = new ResourcePool(builder.maxQueueSize, protocolFactory);
+
+            Set<MetricTag> commonTags = addMetricTagsToSet(builder.commonTags);
+
+            // Set and ensure required tags
+            if (!builder.commonTags.containsKey(SERVICE_TAG)) {
+                if (builder.service == null || builder.service.isEmpty()) {
+                    throw new IllegalArgumentException(String.format("%s common tag is required", SERVICE_TAG));
+                }
+
+                commonTags.add(createMetricTag(SERVICE_TAG, builder.service));
+            }
+            if (!builder.commonTags.containsKey(ENV_TAG)) {
+                if (builder.env == null || builder.env.isEmpty()) {
+                    throw new IllegalArgumentException(String.format("%s common tag is required", ENV_TAG));
+                }
+
+                commonTags.add(createMetricTag(ENV_TAG, builder.env));
+            }
+            if (builder.includeHost) {
+                if (!builder.commonTags.containsKey(HOST_TAG)) {
+                    commonTags.add(createMetricTag(HOST_TAG, InetAddress.getLocalHost().getHostName()));
+                }
+            }
+
+            metricBatch = resourcePool.getMetricBatch();
+            metricBatch.setCommonTags(commonTags);
+            metricBatch.setMetrics(new ArrayList<Metric>());
+
+            calcProtocol = resourcePool.getProtocol();
+            metricBatch.write(calcProtocol);
+            calc = (TCalcTransport) calcProtocol.getTransport();
+            int numOverheadBytes = EMIT_METRIC_BATCH_OVERHEAD + calc.getSizeAndReset();
+
+            freeBytes = builder.maxPacketSizeBytes - numOverheadBytes;
+            if (freeBytes <= 0) {
+                throw new IllegalArgumentException("Common tags serialized size exceeds packet size");
+            }
+
+            bucketIdTagName = builder.histogramBucketIdName;
+            bucketTagName = builder.histogramBucketName;
+            bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
+
+            metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+            executor = builder.executor;
+
+            addAndRunProcessor();
+        } catch (Exception e) {
```

Comment:
This try-catch is going to loose Exception type information. E.g. it'll make IOExceptions (e.g. Thrift connection issues) into IllegalArgumentException. Can you avoid doing that?

Reply:
Because I hand the Exception `e` over to my throw, the exception call stack will be preserved, e.g. it'll say IllegalArgumentException "Exception creating M3Reporter" caused by SocketException "something else".

I agree this isn't the prettiest, but an exception here probably signifies an IllegalArgumentException e.g. an incorrect/already used port. We also need to convert these checked exceptions to unchecked exceptions because we can't recover and proceed from these errors.